go-controller/pkg/nebulaphotomode/nebulaphotomode.go

```diff
@@ -0,0 +1,245 @@
+	return "/sounds/nebulaphotomode.wav"
+}
+
+func (m *NebulaPhotoMode) Start(ctx context.Context) {
+	m.stopWG.Add(1)
+	var loopCtx context.Context
+	loopCtx, m.cancel = context.WithCancel(ctx)
+	go m.loop(loopCtx)
+}
+
+func (m *NebulaPhotoMode) Stop() {
+	m.cancel()
+	m.stopWG.Wait()
+}
+
+func (m *NebulaPhotoMode) loop(ctx context.Context) {
+	defer m.stopWG.Done()
+	defer m.stopSequence()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case event := <-m.joystickEvents:
+			switch event.Type {
+			case joystick.EventTypeButton:
+				if event.Value == 1 {
+					switch event.Number {
+					case joystick.ButtonR1:
+						fmt.Println("NEBULAPHOTO: Getting ready!")
+						m.startWG.Add(1)
+						m.startSequence()
+					case joystick.ButtonSquare:
+						m.stopSequence()
+					case joystick.ButtonTriangle:
+						m.pauseOrResumeSequence()
+					case joystick.ButtonCircle:
+						atomic.StoreInt32(&m.savePicture, 1)
+					}
+				} else {
+					switch event.Number {
+					case joystick.ButtonR1:
+						fmt.Println("NEBULAPHOTO: GO!")
+						m.startWG.Done()
+					}
+				}
+			}
+		}
+	}
+}
+
+func (m *NebulaPhotoMode) startSequence() {
+	if m.running {
+		fmt.Println("NEBULAPHOTO: Already running")
+		return
+	}
+
+	fmt.Println("NEBULAPHOTO: Starting sequence...")
+	m.running = true
+	atomic.StoreInt32(&m.paused, 0)
+
+	seqCtx, cancel := context.WithCancel(context.Background())
+	m.cancelSequence = cancel
+	m.sequenceWG.Add(1)
+	go m.runSequence(seqCtx)
+}
+
+func (m *NebulaPhotoMode) takePicture() (err error) {
+	webcam, werr := gocv.VideoCaptureDevice(0)
+	if werr != nil {
+		err = fmt.Errorf("error opening video capture device: %v", werr)
+		return
+	}
+	defer webcam.Close()
+
+	img := gocv.NewMat()
+	defer img.Close()
+	if ok := webcam.Read(img); !ok {
+		err = fmt.Errorf("cannot read picture from webcam device")
+		return
+	}
+	if img.Empty() {
+		err = fmt.Errorf("no image on device")
+		return
+	}
+	fmt.Printf("NEBULAPHOTO: Read image %v x %v\n", img.Cols(), img.Rows())
+	m.pictureIndex++
+	saveFile := fmt.Sprintf("/tmp/nebula-image-%v.jpg", m.pictureIndex)
+	success := gocv.IMWrite(saveFile, img)
+	fmt.Printf("NEBULAPHOTO: wrote %v? %v\n", saveFile, success)
+	return
+}
+
+func (m *NebulaPhotoMode) fatal(err error) {
+	// Placeholder for what to do if we hit a fatal error.
+	// Callers assume that this does not return normally.
+	panic(err)
+}
+
+// runSequence is a goroutine that reads from the camera and controls the motors.
+func (m *NebulaPhotoMode) runSequence(ctx context.Context) {
+	defer m.sequenceWG.Done()
+	defer fmt.Println("NEBULAPHOTO: Exiting sequence loop")
+	defer m.hw.StopMotorControl()
+
+	// Let the user know that we're ready, then wait for the "GO" signal.
+	m.hw.PlaySound("/sounds/ready.wav")
+	m.startWG.Wait()
+
+	for ii := range m.config.Sequence {
+
+		fmt.Println("NEBULAPHOTO: Next target ball: ", m.config.Sequence[ii])
+		m.announceTargetBall(ii)
+
+		for {
+			time.Sleep(1 * time.Second)
```

Comment:
Is this all that delays the loop?  I think it'll take longer than a second to swap out the targets.  You could probably do away with the loop here and just call takePicture synchronously from a button event.

Reply:
Never mind, I see the atomic stuff now